src/firecracker/src/main.rs

```diff
@@ -397,6 +422,54 @@ fn main_exec() -> Result<(), MainError> {
     }
 }
 
+/// Attempts to resize the processes file descriptor table to match RLIMIT_NOFILE or 2048 if no
+/// RLIMIT_NOFILE is set (this can only happen if firecracker is run outside the jailer. 2048 is
+/// the default the jailer would set).
+///
+/// We do this resizing because the kernel default is 64, with a reallocation happening whenever
+/// the tabel fills up. This was happening for some larger microVMs, and reallocating the
+/// fdtable while a lot of file descriptors are active (due to being eventfds/timerfds registered
+/// to epoll) incurs a penalty of 30ms-70ms on the snapshot restore path.
+fn resize_fdtable() -> Result<(), ResizeFdTableError> {
+    let mut rlimit = libc::rlimit {
+        rlim_cur: 0,
+        rlim_max: 0,
+    };
+
+    // SAFETY: We pass a pointer to a valid area of memory to which we have exclusive mutable access
+    if unsafe { libc::getrlimit(libc::RLIMIT_NOFILE, &mut rlimit as *mut libc::rlimit) } < 0 {
+        return Err(ResizeFdTableError::GetRlimit);
+    }
+
+    // If no jailer is used, there might not be an NOFILE limit set. In this case, resize
+    // the table to the default that the jailer would usually impose (2048)
+    let limit: libc::c_int = if rlimit.rlim_cur == libc::RLIM_INFINITY {
+        2048
+    } else {
+        rlimit.rlim_cur.try_into().unwrap_or(2048)
+    };
+
+    // Resize the file descriptor table to its maximal possible size, to ensure that
+    // firecracker will not need to reallocate it later. If the file descriptor table
+    // needs to be reallocated (which by default happens once more than 64 fds exist,
+    // something that happens for reasonably complex microvms due to each device using
+    // a multitude of eventfds), this can incur a significant performance impact (it
```

Comment:
I started wondering if we want to mention that in the changelog.

Reply:
Yeah, that's fair. I'll get to that after my lunch break

Reply:
Done!